src/parser.rs

```diff
@@ -330,75 +332,123 @@ pub fn video_data_header(input: &[u8]) -> IResult<&[u8], VideoDataHeader> {
     )
   ));
 
-  IResult::Done(&input[1..], VideoDataHeader {
+  assert_eq!(input.len() - 1, remaining.len());
+
+  IResult::Done(remaining, VideoDataHeader {
     frame_type: frame_type,
     codec_id:   codec_id,
   })
 }
 
-#[derive(Debug,PartialEq,Eq)]
+#[derive(Debug, PartialEq)]
 pub struct ScriptDataObject<'a> {
   name: &'a str,
-  data: ScriptDataValue<'a>
+  data: ScriptDataValue<'a>,
 }
 
-#[derive(Debug,PartialEq,Eq)]
+#[derive(Debug, PartialEq)]
+pub struct ScriptDataDate {
+  date_time: f64,
+  local_date_time_offset: i16, // SI16
+}
+
+#[derive(Debug, PartialEq)]
 pub enum ScriptDataValue<'a> {
-  Number,
-  Boolean,
+  Number(f64),
+  Boolean(bool),
   String(&'a str),
-  Object,
-  MovieClip,
+  Object(Vec<ScriptDataObject<'a>>),
+  MovieClip(&'a str),
   Null,
-  UNdefined,
-  Reference,
-  ECMAArray,
-  StrictArray,
-  Date,
+  Undefined,
+  Reference(u16),
+  ECMAArray(Vec<ScriptDataObject<'a>>),
+  StrictArray(Vec<ScriptDataObject<'a>>),
+  Date(ScriptDataDate),
   LongString(&'a str),
 }
 
-/*
 named!(pub script_data_object<ScriptDataObject>,
-  chain!(
-    name: script_data_string ~
-    data: script_data_value  ,
-    || {
-      ScriptDataObject {
+  do_parse!(
+    name: script_data_string >>
+    data: get_data_value     >>
+    (ScriptDataObject {
         name: name,
-        data: data
-      }
-    }
+        data: data,
+    })
   )
 );
 
-pub fn script_data_object_end(input:&[u8]) -> IResult<&[u8],()> {
-  match be_u24(input) {
-    IResult::Done(i,o) => if o == 9 {
-      IResult::Done(i,())
-    } else {
-      IResult::Error(Err::Code(ErrorKind::Tag))
-    },
-    e => e
-  }
+
+pub fn script_data_object_end(input:&[u8]) -> IResult<&[u8],&[u8]> {
+  static tag: &'static [u8] = &[0u8, 0u8, 9u8];
+  tag!(tag, input)
```

Comment:
it should be `tag!(input, tag)` here. That's probably why `terminated!` appeared to fail